Movie update graphql

Mi4-04 opened the pull request.

Mi4-04 committed:
add graphQL in project

Mi4-04 committed:
Merge branch 'master' of https://github.com/Mi4-04/movies-front into movie-update-graphql

Mi4-04 committed:
add mutation and query in app

-add jwt auth
-add update watched
-add remove film
-add list movies

reshetov commented on src/Pages/AddPage/component/MoviesList/index.tsx:
```diff
@@ -24,31 +27,50 @@ export const MoviesList = ({
+    });
   };
 
+  const { data, loading } = useQuery(GET_ALL_MOVIES, {
```
Не знаю что у тебя на беке, но так работать не будет. API отдает фильмы постранично а тут у тебя нет page от слова совсем. Либо я что-то не знаю, либо ты работаешь только с первой страницей фильмов.

reshetov commented on src/Pages/Main/component/FavoriteMovies/index.tsx:
```diff
@@ -23,19 +29,45 @@ export const FavoriteMovies = ({ blockView }: IFavoriteMoviesProps) => {
 
+  let movieId;
+
+  let newFilmsIds = movies.map((film) => (movieId = film.id));
```
Чет тут какая-то странная логика, у тебя после этого прогона в movieId будет ПОСЛЕДНИЙ id из movies.
Если тебе нужен какой-то определенный - то для этого filter есть, или я что-то не понимаю

reshetov commented on src/Pages/Main/component/FavoriteMovies/index.tsx:
```diff
@@ -23,19 +29,45 @@ export const FavoriteMovies = ({ blockView }: IFavoriteMoviesProps) => {
     movies[index].watched = !movies[index].watched;
     setMovies([...movies]);
+    let filmId;
+    movies.map((movie) => (filmId = movie.id));
```
И тут тоже

reshetov commented on src/gql/mutation/index.ts:
```diff
@@ -30,19 +32,23 @@ export const ADD_FAV_MOVIES = gql`
-  ${CORE_MOVIE_FIELDS}
   mutation UpdateWatched($updateWatchedId: Float!) {
     updateWatched(id: $updateWatchedId) {
-      ...CoreMovieFields
```
А зачем ты убрал фрагмент?

reshetov commented on src/App.tsx:
```diff
@@ -48,47 +47,42 @@ const App = () => {
+
+            <Switch>
+              <Route exact path="/">
+                <Main
```
Если уж ты решил тут сразу и авторизацию припилить, то этот роутинг должен быть приватным. Без авторизации только логин доступен

reshetov commented on src/App.tsx:
```diff
@@ -48,47 +47,42 @@ const App = () => {
+              <AuthRoute exact path="/login">
+                <LoginPage setUserLogin={setUserLogin} />
+              </AuthRoute>
+              <Route path="/add">
```
И этот тоже

Isidora25 commented on src/utils/user.tsx:
```diff
@@ -0,0 +1,50 @@
+
+const AuthProvider = (props: any) => {
+  const jwtToken = localStorage.getItem("accessToken");
+  const [users, setUser] = useState(jwtToken);
```
users?

Isidora25 commented on src/utils/user.tsx:
```diff
@@ -0,0 +1,50 @@
+import jwtDecode from "jwt-decode";
+
+interface AuthContextState {
+  user: null;
```
А что за тип такой, `null`?

Isidora25 commented on src/utils/user.tsx:
```diff
@@ -0,0 +1,50 @@
+
+interface AuthContextState {
+  user: null;
+  signIn: (data: string) => void;
```
data? какая data? 
Ты сам не знаешь, что туда передаешь? 
Через месяц сам не вспомнишь, что за data
Назови нормально, чтобы не одному тебе было понятно, что передается в качестве аргументов

Isidora25 commented on src/utils/user.tsx:
```diff
@@ -0,0 +1,50 @@
+  const signIn = useCallback(
+    (token) => {
+      localStorage.setItem("accessToken", token);
+      setUser(token);
```
почему в user хранится только token?

Isidora25 commented on src/utils/user.tsx:
```diff
@@ -0,0 +1,50 @@
+  const logout = useCallback(() => {
+    localStorage.removeItem("accessToken");
+    localStorage.removeItem("userLogin");
+    setUser(null);
```
Если не ошибаюсь, тут еще и кеш чистить нужно 
Проверь документацию аполло клиента

Isidora25 commented on src/Pages/Login/index.tsx:
```diff
@@ -1,32 +1,52 @@
-    if (isValidUser(data.login, data.password)) {
-      return { [FORM_ERROR]: t("authorization.incorrectData") };
-    }
+  const context = React.useContext(AuthContext);
```
почему бы не сделать деструктуризацию и сразу получить нужные данные из контекста?

Isidora25 replied:
Почему в одном месте в проекте useContext импортируется из библиотеки React, а здесь иначе? 
Придерживайся единого стиля в коде, это важно

Isidora25 commented on src/Pages/Login/index.tsx:
```diff
@@ -1,32 +1,52 @@
+    return <CircularProgress color="success" />;
+  }
+
+  const onSubmit = (data: ILogin) => {
```
используй деструктуризацию

Isidora25 commented on src/Pages/Login/index.tsx:
```diff
@@ -1,32 +1,52 @@
+        password: data.password,
+      },
+      onCompleted: (data) => {
+        context.signIn(data.signIn.accessToken);
```
Вот я тут не очень понимаю, почему signIn - это объект теперь?
У тебя это описано как функция в контексте

Isidora25 commented on src/Pages/Login/index.tsx:
```diff
@@ -1,32 +1,52 @@
+        login: data.login,
+        password: data.password,
+      },
+      onCompleted: (data) => {
```
опять data 
В одной функции две переменных  с именем data 
Очень не информативно и серьезно сбивает с толку.

Isidora25 commented on src/components/PrivateRouter/index.tsx:
```diff
@@ -0,0 +1,17 @@
+    <Route
+      {...args}
+      render={(props) =>
+        !user ? <Component {...props} /> : <Redirect to="/" />
```
Вообще не понимаю смысл этого кастомного роута 
Auth - это какой? приватный или публичный? 
Логично было бы предположить, что ты хотел сделать приватный роут. 
Но почему тогда ты рендеришь компонент, если юзер не определен? 
Получается, ты отправляешь на статичный роут, если юзер залогинен? 

очень странная логика

Isidora25 commented on src/index.tsx:
```diff
@@ -1,20 +1,12 @@
-      <App />
-    </Suspense>
-  </ApolloProvider>,
+  <Suspense fallback={<div>Loading...</div>}>
```
Я надеюсь, что ты используешь экспериментальный функционал полностью принимая всю ответственность и понимая, что это может тебе выстрелить в ногу

Mi4-04 committed:
fix comments

-add fragment gql
-add pagination
-more fix comment

Isidora25 commented on src/Pages/AddPage/component/MoviesList/index.tsx:
```diff
@@ -22,11 +23,18 @@ export const MoviesList = ({
   );
 
+  const handleChangePage = (
+    event: React.ChangeEvent<unknown>,
```
Если тебе приходится объявлять параметр функции, но в использовании он тебе не нужен, используй знак нижнего подчеркивания _

Isidora25 commented on src/utils/user.tsx:
```diff
@@ -1,49 +1,56 @@
 interface AuthContextState {
-  user: null;
-  signIn: (data: string) => void;
+  token: any;
```
какие типы тут могут быть?

Isidora25 commented on src/utils/user.tsx:
```diff
@@ -1,49 +1,56 @@
-  user: null;
-  signIn: (data: string) => void;
+  token: any;
+  signIn: (token: string) => void;
```
а здесь token:string

Mi4-04 committed:
fix type AuthContext and handleChangePage

Mi4-04 committed:
fix rename to backend

reshetov commented on src/Pages/Main/component/FavoriteMovies/index.tsx:
```diff
@@ -23,19 +29,45 @@ export const FavoriteMovies = ({ blockView }: IFavoriteMoviesProps) => {
 
+  let movieId;
+
+  let newFilmsIds = movies.filter((film) => (movieId = film.id));
```
Как я вижу эту логику, ты объявляешь переменную movieId и потом в итераторе filter пробегаешься по всем фильмам и каждый раз присваиваешь этой переменной значение id от фильма.
Итого: у тебя после этого перебора в movieID будет id ПОСЛЕДНЕГО фильма из массива.
Вопрос: А зачем все это? И если уж тебе вот прям зачем-то нужен тот ID можно сразу его по индексу взять. Но больше это на какой-то костыль похоже

reshetov commented on src/Pages/Main/component/FavoriteMovies/index.tsx:
```diff
@@ -23,19 +29,45 @@ export const FavoriteMovies = ({ blockView }: IFavoriteMoviesProps) => {
     movies[index].watched = !movies[index].watched;
     setMovies([...movies]);
+    let filmId;
+    movies.filter((movie) => (filmId = movie.id));
```
См. коммент выше

reshetov commented on src/config/apolloClient.ts:
```diff
@@ -0,0 +1,25 @@
+
+import { setContext } from "@apollo/client/link/context";
+
+const httpLink = createHttpLink({ uri: "http://localhost:4000/graphql" });
```
Вынеси в URL в переменную окружения

Mi4-04 committed:
fix favorite movies

reshetov commented on src/Pages/Main/component/FavoriteMovies/index.tsx:
```diff
@@ -23,19 +29,43 @@ export const FavoriteMovies = ({ blockView }: IFavoriteMoviesProps) => {
     });
   }, []);
 
+  let newFilmsIds = movies.map((film) => film.id);
```
Ну допустим теперь ты собираешь массив из id. Вопрос остается тем-же - а зачем? Для чего вообще хранить ids фильмов что тебе вернул API?

reshetov commented on src/Pages/Main/component/FavoriteMovies/index.tsx:
```diff
@@ -23,19 +29,43 @@ export const FavoriteMovies = ({ blockView }: IFavoriteMoviesProps) => {
 
+  let newFilmsIds = movies.map((film) => film.id);
+
+  let newFilms = movies.find((film) => film.id);
```
1. film(s) - Это множественное число, а find отдает один объект
2. Подумай, а точно тебе нужна это строка?

Mi4-04 committed:
fix comment
-add updateWatched
-add getFavMovies
-fix favoriteMoviesItem
-fix year select type number

Mi4-04 committed:
change FavoriteMoviesItem

-add localstorage for storage state watched
-add dectrutarication component

Isidora25 commented on src/Pages/Main/component/FavoriteMovies/component/FavoriteMoviesItem/index.tsx:
```diff
@@ -14,15 +14,21 @@ import React from "react";
 
   const [updateWatched] = useMutation(UPDATE_WATCHED);
-  const [watched, setWatched] = React.useState<boolean>(false);
+  const [watched, setWatched] = React.useState<any>(
```
any?

Isidora25 commented on src/Pages/Main/component/FavoriteMovies/component/FavoriteMoviesItem/index.tsx:
```diff
@@ -14,15 +14,21 @@ import React from "react";
   const [updateWatched] = useMutation(UPDATE_WATCHED);
-  const [watched, setWatched] = React.useState<boolean>(false);
+  const [watched, setWatched] = React.useState<any>(
+    JSON.parse(String(localStorage.getItem(String(film.id))))
```
Что-то странное тут происходит

watched - судя по названию, boolean 
localStorage.getItem - возвращает string  или null 

для чего JSON.parse?

Isidora25 commented on src/Pages/Main/component/FavoriteMovies/component/FavoriteMoviesItem/index.tsx:
```diff
@@ -14,15 +14,21 @@ import React from "react";
+}
+
+export const FavoriteMoviesItem = ({
+  film,
```
Используй сразу деструктуризацию, чтобы не тащить везде film

Mi4-04 committed:
fix coment

-fix type useState
-add distructarisation film
